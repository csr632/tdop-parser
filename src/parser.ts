import type { Scanner } from "./scanner";

export function createParser(scanner: Scanner) {
  const parser = {
    parseExp,
  };

  function parseExp(ctxPrecedence: number): any {
    let prefixToken = scanner.consume();
    if (!prefixToken) throw new Error();

    const prefixParselet =
      prefixParselets[prefixToken] ?? prefixParselets.__default;
    let left = prefixParselet.handle(prefixToken, parser);

    let infixToken = scanner.peek();
    while (infixToken) {
      const infixParselet = infixParselets[infixToken];
      if (!infixParselet) throw new Error(infixToken);
      if (infixParselet.precedence <= ctxPrecedence) break;
      scanner.consume();
      left = infixParselet.handle(left, infixToken, parser);
      infixToken = scanner.peek();
    }
    return left;
  }

  return function parse() {
    return parseExp(0);
  };
}

interface PrefixParseLet {
  handle(token: string, parser: any): any;
}
const prefixParselets: Record<string, PrefixParseLet> = {
  __default: {
    handle(token, parser) {
      return {
        type: "value",
        value: token,
      };
    },
  },
};
function registerPrefixOperator(prefix: string, precedence: number) {
  prefixParselets[prefix] = {
    handle(token, { parseExp }) {
      const body = parseExp(precedence);
      if (!body) throw new Error(`invalid prefix usage "${prefix}"`);
      return {
        type: `prefix${prefix}`,
        body,
      };
    },
  };
}
registerPrefixOperator("+", 50);
registerPrefixOperator("-", 50);

interface InfixParseLet {
  handle(left: any, token: string, parser: any): any;
  precedence: number;
}
const infixParselets: Record<string, InfixParseLet> = {};
function registerInfixOperator(
  infix: string,
  precedence: number,
  associateRight2Left: boolean = false
) {
  infixParselets[infix] = {
    precedence,
    handle(left, token, { parseExp }) {
      const right = parseExp(associateRight2Left ? precedence - 1 : precedence);
      if (!right) throw new Error(`invalid infix usage "${infix}"`);
      return {
        type: `infix${infix}`,
        left,
        right,
      };
    },
  };
}
registerInfixOperator("+", 10);
registerInfixOperator("-", 10);
registerInfixOperator("*", 20);
registerInfixOperator("/", 20);
