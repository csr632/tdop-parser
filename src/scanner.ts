export interface Scanner {
  peek(): string | null;
  consume(): string | null;
}

export function createScanner(text: string): Scanner {
  const iterator = text[Symbol.iterator]();
  let next = iterator.next();
  let peeked = false;
  let token: string | null = null;

  return {
    peek(): string | null {
      if (!peeked) token = scanNextToken();
      peeked = true;
      return token;
    },
    consume(): string | null {
      if (peeked) {
        peeked = false;
        return token;
      }
      token = scanNextToken();
      return token;
    },
  };

  function scanNextToken(): string | null {
    // skip WhiteSpace at front
    while (!next.done && isWhiteSpace(next.value)) {
      next = iterator.next();
    }

    if (next.done) return null;

    if (isTokenBoundary(next.value)) {
      const token = next.value;
      next = iterator.next();
      return token;
    }

    let token = "";
    while (!next.done && !isTokenBoundary(next.value)) {
      token += next.value;
      next = iterator.next();
    }
    return token;
  }
}

// token is a sequence of \w
function isTokenBoundary(char: string) {
  if (!char) debugger;
  return !char.match(/^\w$/);
}

function isWhiteSpace(char: string) {
  return char.match(/\s/);
}
