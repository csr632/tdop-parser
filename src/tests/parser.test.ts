import { expect, test } from "vitest";
import { createParser } from "../parser";
import { createScanner } from "../scanner";

test("parse -1+2*6/3+5", () => {
  const parser = createParser(createScanner("-1+2*6/3+5"));
  const res = parser.parseProgram();
  expect(res).toMatchSnapshot();
});

test("parse 1 + 2^3^4*5-6", () => {
  const parser = createParser(createScanner("1 + 2^3^4*5-6"));
  const res = parser.parseProgram();
  expect(res).toMatchSnapshot();
  // console.log(JSON.stringify(res, null, 2))
});

test("parse true ? 1 : 2", () => {
  const parser = createParser(createScanner("true ? 1 : 2"));
  const res = parser.parseProgram();
  expect(res).toMatchSnapshot();
});

test("parse 1+ 2  *3 ? 1+1 : 2  -1", () => {
  const parser = createParser(createScanner("1+ 2  *3 ? 1+1 : 2  -1"));
  const res = parser.parseProgram();
  expect(res).toMatchSnapshot();
});

test("parse conditional chains", () => {
  const parser = createParser(
    createScanner(`
  condition1 ? value1
  : condition2 ? value2
  : condition3 ? value3
  : value4`)
  );
  const res = parser.parseProgram();
  expect(res).toMatchSnapshot();
});
test("parse conditional chains", () => {
  const parser = createParser(
    createScanner(`
  condition1
    ? condition2 ? true2 : false2
    : condition3
      ? condition4 ? true4 : false4
      : condition5 ? true5 : false5
  `)
  );
  const res = parser.parseProgram();
  expect(res).toMatchSnapshot();
});

test("parse true ? 1 ; 2", () => {
  expect(() => {
    const parser = createParser(createScanner("true ? 1 ; 2"));
    parser.parseProgram();
  }).toThrowError("expect token : but got ;");
});

test("parse 1 + ", () => {
  expect(() => {
    const parser = createParser(createScanner("1 + "));
    parser.parseProgram();
  }).toThrowError("expect token but found none");
});

test("parse 1 2", () => {
  const parser = createParser(createScanner("1 2"));
  const res = parser.parseProgram();
  expect(res).toMatchSnapshot();
});

test("parse 2*(3-1)", () => {
  const parser = createParser(createScanner("2*(3-1)"));
  const res = parser.parseProgram();
  expect(res).toMatchSnapshot();
});

test("parse ((1))+1", () => {
  const parser = createParser(createScanner("((1))+1"));
  const res = parser.parseProgram();
  expect(res).toMatchSnapshot();
});

test("parse 2*(9*(((2))+1))-1", () => {
  const parser = createParser(createScanner("2*(9*(((2))+1))-1"));
  const res = parser.parseProgram();
  expect(res).toMatchSnapshot();
});

test("parse 2*(3 1)", () => {
  expect(() => {
    const parser = createParser(createScanner("2*(3 1)"));
    parser.parseProgram();
  }).toThrowError("expect token ) but got 1");
});
