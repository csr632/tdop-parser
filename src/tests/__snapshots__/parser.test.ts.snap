// Vitest Snapshot v1

exports[`basic > -1+2*6/3+5 1`] = `
{
  "left": {
    "left": {
      "operator": "-",
      "right": {
        "type": "value",
        "value": "1",
      },
      "type": "unary",
    },
    "operator": "+",
    "right": {
      "left": {
        "left": {
          "type": "value",
          "value": "2",
        },
        "operator": "*",
        "right": {
          "type": "value",
          "value": "6",
        },
        "type": "binary",
      },
      "operator": "/",
      "right": {
        "type": "value",
        "value": "3",
      },
      "type": "binary",
    },
    "type": "binary",
  },
  "operator": "+",
  "right": {
    "type": "value",
    "value": "5",
  },
  "type": "binary",
}
`;

exports[`basic > 1 + 2^3^4*5-6 1`] = `
{
  "left": {
    "left": {
      "type": "value",
      "value": "1",
    },
    "operator": "+",
    "right": {
      "left": {
        "left": {
          "type": "value",
          "value": "2",
        },
        "operator": "^",
        "right": {
          "left": {
            "type": "value",
            "value": "3",
          },
          "operator": "^",
          "right": {
            "type": "value",
            "value": "4",
          },
          "type": "binary",
        },
        "type": "binary",
      },
      "operator": "*",
      "right": {
        "type": "value",
        "value": "5",
      },
      "type": "binary",
    },
    "type": "binary",
  },
  "operator": "-",
  "right": {
    "type": "value",
    "value": "6",
  },
  "type": "binary",
}
`;

exports[`conditional operation > 1+ 2  *3 ? 1+1 : 2  -1 1`] = `
{
  "condition": {
    "left": {
      "type": "value",
      "value": "1",
    },
    "operator": "+",
    "right": {
      "left": {
        "type": "value",
        "value": "2",
      },
      "operator": "*",
      "right": {
        "type": "value",
        "value": "3",
      },
      "type": "binary",
    },
    "type": "binary",
  },
  "falseBranch": {
    "left": {
      "type": "value",
      "value": "2",
    },
    "operator": "-",
    "right": {
      "type": "value",
      "value": "1",
    },
    "type": "binary",
  },
  "trueBranch": {
    "left": {
      "type": "value",
      "value": "1",
    },
    "operator": "+",
    "right": {
      "type": "value",
      "value": "1",
    },
    "type": "binary",
  },
  "type": "conditional",
}
`;

exports[`conditional operation > conditional chains 1`] = `
{
  "condition": {
    "type": "value",
    "value": "condition1",
  },
  "falseBranch": {
    "condition": {
      "type": "value",
      "value": "condition2",
    },
    "falseBranch": {
      "condition": {
        "type": "value",
        "value": "condition3",
      },
      "falseBranch": {
        "type": "value",
        "value": "value4",
      },
      "trueBranch": {
        "type": "value",
        "value": "value3",
      },
      "type": "conditional",
    },
    "trueBranch": {
      "type": "value",
      "value": "value2",
    },
    "type": "conditional",
  },
  "trueBranch": {
    "type": "value",
    "value": "value1",
  },
  "type": "conditional",
}
`;

exports[`conditional operation > conditional chains 2`] = `
{
  "condition": {
    "type": "value",
    "value": "condition1",
  },
  "falseBranch": {
    "condition": {
      "type": "value",
      "value": "condition3",
    },
    "falseBranch": {
      "condition": {
        "type": "value",
        "value": "condition5",
      },
      "falseBranch": {
        "type": "value",
        "value": "false5",
      },
      "trueBranch": {
        "type": "value",
        "value": "true5",
      },
      "type": "conditional",
    },
    "trueBranch": {
      "condition": {
        "type": "value",
        "value": "condition4",
      },
      "falseBranch": {
        "type": "value",
        "value": "false4",
      },
      "trueBranch": {
        "type": "value",
        "value": "true4",
      },
      "type": "conditional",
    },
    "type": "conditional",
  },
  "trueBranch": {
    "condition": {
      "type": "value",
      "value": "condition2",
    },
    "falseBranch": {
      "type": "value",
      "value": "false2",
    },
    "trueBranch": {
      "type": "value",
      "value": "true2",
    },
    "type": "conditional",
  },
  "type": "conditional",
}
`;

exports[`conditional operation > true ? 1 : 2 1`] = `
{
  "condition": {
    "type": "value",
    "value": "true",
  },
  "falseBranch": {
    "type": "value",
    "value": "2",
  },
  "trueBranch": {
    "type": "value",
    "value": "1",
  },
  "type": "conditional",
}
`;

exports[`error handling > 1 2 1`] = `
{
  "type": "value",
  "value": "1",
}
`;

exports[`parenthesis > ((1))+1 1`] = `
{
  "left": {
    "type": "value",
    "value": "1",
  },
  "operator": "+",
  "right": {
    "type": "value",
    "value": "1",
  },
  "type": "binary",
}
`;

exports[`parenthesis > 2*(3-1) 1`] = `
{
  "left": {
    "type": "value",
    "value": "2",
  },
  "operator": "*",
  "right": {
    "left": {
      "type": "value",
      "value": "3",
    },
    "operator": "-",
    "right": {
      "type": "value",
      "value": "1",
    },
    "type": "binary",
  },
  "type": "binary",
}
`;

exports[`parenthesis > 2*(9*(((2))+1))-1 1`] = `
{
  "left": {
    "left": {
      "type": "value",
      "value": "2",
    },
    "operator": "*",
    "right": {
      "left": {
        "type": "value",
        "value": "9",
      },
      "operator": "*",
      "right": {
        "left": {
          "type": "value",
          "value": "2",
        },
        "operator": "+",
        "right": {
          "type": "value",
          "value": "1",
        },
        "type": "binary",
      },
      "type": "binary",
    },
    "type": "binary",
  },
  "operator": "-",
  "right": {
    "type": "value",
    "value": "1",
  },
  "type": "binary",
}
`;
